BaseClasses.py

```diff
@@ -759,20 +742,117 @@ def sweep_for_events(self, locations: Optional[Iterable[Location]] = None) -> No
+
+            for player, locations in advancements_per_player:
+                if player not in players_to_check:
+                    next_advancements_per_player.append((player, locations))
+                    continue
+
+                # Accessibility of each location is checked first because a player's region accessibility cache becomes
+                # stale whenever one of their own items is collected into the state.
+                reachable_locations: List[Location] = []
+                unreachable_locations: List[Location] = []
+                for location in locations:
+                    if location.can_reach(self):
+                        # Locations containing items that do not belong to `player` could be collected immediately
+                        # because they won't stale `player`'s region accessibility cache, but, for simplicity, all the
+                        # items at reachable locations are collected in a single loop.
+                        reachable_locations.append(location)
+                    else:
+                        unreachable_locations.append(location)
+                if unreachable_locations:
+                    next_advancements_per_player.append((player, unreachable_locations))
+
+                # If the player does not receive any items for the rest of the sweep iteration, they will likely not
+                # have any reachable locations at the start of the next sweep iteration, so their locations can be
+                # skipped.
+                next_players_to_check.discard(player)
+
+                # Collect the items from the reachable locations.
+                for advancement in reachable_locations:
+                    self.advancements.add(advancement)
+                    item = advancement.item
+                    assert isinstance(item, Item), "tried to collect advancement Location with no Item"
+                    if self.collect(item, True, advancement):
+                        # The player the item belongs to may be able to reach additional locations in the next sweep
+                        # iteration.
+                        next_players_to_check.add(item.player)
+
+            if not next_players_to_check:
+                if not checking_if_finished:
+                    # It is assumed that each player's world only logically depends on itself, which may not be the
+                    # case, so confirm that the sweep is finished by doing an extra iteration that checks every player.
+                    checking_if_finished = True
+                    next_players_to_check = all_players
+            else:
+                checking_if_finished = False
+
+            players_to_check = next_players_to_check
+            advancements_per_player = next_advancements_per_player
+
+            if yield_each_sweep:
+                yield
+
+    def sweep_for_advancements(self, locations: Optional[Iterable[Location]] = None, yield_each_sweep: bool = False,
```

Comment:
```suggestion
                yield

    @typing.overload
    def sweep_for_advancements(self, locations: Optional[Iterable[Location]] = None, *,
                               yield_each_sweep: typing.Literal[True],
                               checked_locations: Optional[Set[Location]] = None) -> Iterator[None]: ...

    @typing.overload
    def sweep_for_advancements(self, locations: Optional[Iterable[Location]] = None,
                               yield_each_sweep: typing.Literal[False] = False,
                               checked_locations: Optional[Set[Location]] = None) -> None: ...

    def sweep_for_advancements(self, locations: Optional[Iterable[Location]] = None, yield_each_sweep: bool = False,
```
This makes it so type-checkers and IDEs can see which calls return `None` and which calls return `Iterator`
so that it doesn't complain about returning an `Iterator` from `sweep_for_events`
or about iterating through `None` in `can_beat_game`

Reply:
Thanks, I couldn't remember how to get the `yield_each_sweep: typing.Literal[True],` overload to work properly due to the default value of `False`.